File: tests/backports/host/scenekit/inventory.swift
// inventory <scenes-dir> : what macOS SceneKit finds in each of Telegram's ten scenes (facts/SceneKit/SCNView.md, "Which
// scenes Telegram draws"): the decompressed files star2.scn, coin.scn, gift2.scn, badge.scn, emoji.scn, tag.scn, business.scn,
// boost.scn, lightspeed.scn and swirl.scn (`git show release-12.9.2:submodules/PremiumUI/Resources/<name> | gunzip`).
import SceneKit
import AppKit

let directory = CommandLine.arguments[1]
for name in ["star2", "coin", "gift2", "badge", "emoji", "tag", "business", "boost", "lightspeed", "swirl"] {
    guard let scene = try? SCNScene(url: URL(fileURLWithPath: "\(directory)/\(name).scn"), options: nil) else { print(name, "does not load"); continue }
    var nodes = 0, geometries = 0, lights: [String] = [], models = Set<String>(), particles = 0, subdivided = 0, cameras = 0, imageSlots = 0
    func walk(_ node: SCNNode) {
        nodes += 1
        if let light = node.light { lights.append(light.type.rawValue) }
        if node.camera != nil { cameras += 1 }
        particles += node.particleSystems?.count ?? 0
        if let geometry = node.geometry {
            geometries += 1
            if geometry.subdivisionLevel > 0 { subdivided += 1 }
            for material in geometry.materials {
                models.insert(material.lightingModel.rawValue.replacingOccurrences(of: "SCNLightingModel", with: ""))
                for property in [material.diffuse, material.emission, material.metalness, material.roughness, material.selfIllumination, material.transparent, material.multiply, material.reflective]
                where property.contents is NSImage || property.contents is URL || property.contents is String { imageSlots += 1 }
            }
        }
        node.childNodes.forEach(walk)
    }
    walk(scene.rootNode)
    print(name, "nodes", nodes, "geometries", geometries, "cameras", cameras, "lights", lights.sorted(), "models", models.sorted(), "particleSystems", particles, "subdivided", subdivided, "imageSlots", imageSlots)
}

File: tests/backports/host/scenekit/inventory.py
"""inventory.py <scenes-dir> : which of rotation, orientation and eulerAngles each node of Telegram's ten scenes archives
(facts/SceneKit/SCNView.md, "Conventions"), and, for a node with a rotation and no orientation, that rotation beside its
eulerAngles. <scenes-dir> holds the decompressed files of inventory.swift."""
import plistlib, struct, sys

NAMES = ["star2", "coin", "gift2", "badge", "emoji", "tag", "business", "boost", "lightspeed", "swirl"]
for name in NAMES:
    objects = plistlib.load(open("%s/%s.scn" % (sys.argv[1], name), "rb"))["$objects"]
    nodes = [o for o in objects if isinstance(o, dict) and ("childNodes" in o or "camera" in o or "light" in o or "geometry" in o or ("name" in o and "movabilityHint" in o))]
    both = sum(1 for o in nodes if "orientation" in o and "rotation" in o)
    euler_only = sum(1 for o in nodes if "eulerAngles" in o and "orientation" not in o)
    rotation_only = sum(1 for o in nodes if "rotation" in o and "orientation" not in o)
    print(name, "nodes", len(nodes), "rotation and orientation", both, "eulerAngles without orientation", euler_only, "rotation without orientation", rotation_only)
    for o in nodes:
        if "rotation" in o and "orientation" not in o:
            def floats(key):
                value = o.get(key)
                value = objects[value.data] if isinstance(value, plistlib.UID) else value
                return struct.unpack("<%df" % (len(value) // 4), value) if isinstance(value, bytes) else None
            print("   rotation", floats("rotation"), "eulerAngles", floats("eulerAngles"))
